### ISSUE TITLE
Non End-to-end Optical Music Recognition

### ISSUE THREAD

**Issue by petarbratic** [ftn-ai-lab/sc-2025#29]
**1. Tim**
Petar Bratić RA 14/2022

**2. Asistent**
Dragan Vidaković

**3. Definicija problema**
Optičko prepoznavanje notnih zapisa sa slike (Optical Music Recognition - OMR) primenom tehnika dubokog učenja. Cilj je razvijanje modela koji je u stanju da na osnovu ulazne slike generiše odgovarajući sekvencijalni tekstualni zapis (u obliku niza unapred definisanih tokena).
Ograničiću se na prepoznavanje skeniranih notnih zapisa koji sadrže jedan linijski sistem po slici.

**4. Motivacija problema**
Konvertovanje odštampanih notnih zapisa u digitalni format zahteva angažovanje stručnih lica koja ručno prepisuju muzički sadržaj. Ne samo što je taj proces podložan greškama, već je i veoma skup jer zahteva rad osoba sa odgovarajućim znanjem iz oblasti muzike. Automatizacija ovog procesa bi omogućila značajno ubrzanje procesa digitalizacije i smanjenje troškova. Tada bi se uloga stručnjaka svela na proveru i eventualnu korekciju automatski generisanih zapisa.

**5. Skup podataka**
Za potrebe projekta će se koristiti javno dostupan PrIMuS (Printed Images of Music Staves) skup podataka. Svaka instanca u skupu podataka sadrži sliku notnog zapisa i odgovarajuće sekvence muzičkih tokena. Planirana je podela na trening, validacioni i test skup.

**6. Metodologija**
Arhitektura sistema će biti zasnovana na onoj iz naučnog rada "Musical Note Position and Duration Recognition Model in Optical Music Recognition Using Convolutional Neural Network" (Link 1). 

Prvo će se izvršiti detekcija pojedinačnih nota i izdvajanje regiona od interesa (ROI). Svaki region će biti prosleđen ka dva odvojena konvoluciona klasifikatora:
   - Pitch Recognition (PR) model za određivanje pozicije note u odnosu na notni sistem, tj. određivanje *tona*.
   - Duration Recognition (DR) model za određivanje trajanja note, tj. *notne vrednosti*.

Oba modela biće implementirana kao konvolucione neuronske mreže sa četiri konvoluciona sloja (uz max-pooling nakon svakog sloja), nakon čega sledi jedan potpuno povezan sloj i završni Softmax sloj. Treniranje modela vršiće se primenom categorical cross-entropy funkcije gubitka uz optimizator tipa Adam.

**7. Evaluacija**
Evaluacija će se vršiti na dva nivoa.
1. Pojedinačni klasifikatori (PR i DR) biće evaluirani pomoću standardnih metrika višeklasne klasifikacije (accuracy i precision).
2. Kompletan sistem biće evaluiran na nivou rekonstrukcije notnog zapisa uz pomoć Character Error Rate (CER) i Sequence Error Rate (SER) koje mere odstupanje predviđene sekvence tokena od tačne sekvence.

**8. Linkovi do radova korišćenih u istaživanju teme**

"Musical Note Position and Duration Recognition Model in Optical Music Recognition Using Convolutional Neural Network", Douglas Rakasiwi Nugroho i Amalia Zahra
**Link 1:** https://www.joig.net/2024/JOIG-V12N1-32.pdf

"Optical Music Recognition using Deep Neural Networks", Jiri Mayer
**Link 2:** https://dspace.cuni.cz/bitstream/handle/20.500.11956/119393/130284906.pdf?sequence=1&isAllowed=y

"Deep Learning for Optical Music Recognition: A Review", Francisco J. Castellanos, Antonio Javier Gallego i Ichiro Fujinaga
**Link 3:** https://www.techrxiv.org/users/895830/articles/1272264-deep-learning-for-optical-music-recognition-a-review

**Comment by vdragan1993** [comment]
Skup podataka i metodologija su identični odbranjenom [projektu](https://github.com/sc-2020-siit/MusicSheetsRecognition) koji je demonstriran na terminu vežbi.
Predložite drugačiju metodologiju.

**Comment by petarbratic** [comment]
Poštovani, u skladu sa sugestijom, izmenio sam metodologiju rada.

**Comment by vdragan1993** [comment]
Tema odobrena.
Ostavite link ka GitHub repozitorijumu projekta.
Srećan rad.

**Comment by petarbratic** [comment]
Hvala.

Link: https://github.com/petarbratic/omr
